src/rngs/xoshiro256plusplus.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[cfg(feature="serde1")] use serde::{Serialize, Deserialize};
+use rand_core::impls::fill_bytes_via_next;
+use rand_core::le::read_u64_into;
+use rand_core::{SeedableRng, RngCore, Error};
+
+/// A xoshiro256** random number generator.
+///
+/// The xoshiro256** algorithm is not suitable for cryptographic purposes, but
+/// is very fast and has excellent statistical properties.
+///
+/// The algorithm used here is translated from [the `xoshiro256plusplus.c`
+/// reference source code](http://xoshiro.di.unimi.it/xoshiro256plusplus.c) by
+/// David Blackman and Sebastiano Vigna.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature="serde1", derive(Serialize, Deserialize))]
+pub struct Xoshiro256PlusPlus {
+    s: [u64; 4],
+}
+
+impl SeedableRng for Xoshiro256PlusPlus {
+    type Seed = [u8; 32];
+
+    /// Create a new `Xoshiro256PlusPlus`.  If `seed` is entirely 0, it will be
+    /// mapped to a different seed.
+    #[inline]
+    fn from_seed(seed: [u8; 32]) -> Xoshiro256PlusPlus {
+        if seed.iter().all(|&x| x == 0) {
+            return Self::seed_from_u64(0);
+        }
+        let mut state = [0; 4];
+        read_u64_into(&seed, &mut state);
+        Xoshiro256PlusPlus { s: state }
+    }
+}
+
+impl RngCore for Xoshiro256PlusPlus {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        self.next_u64() as u32
+    }
```

Comment:
This takes the low-order bytes. @vigna is it preferable to take the high-order bytes given that we're discarding half anyway? Or is there no point caring for the ++ variant?

Reply:
Well, I'd take the high bits (longer carry chain propagation, which means higher linear complexity), but, really, the difference in quality is not detectable.

Note that if you use PCG32 to initialize the state usinga  64-bit seed, seeds with a large number of lower bits in common will initialize the state in a similar way (try with 0 and 2^63), and consequently the very first outputs will be correlated. The effect might be however very mild in practice.

You can avoid this problem by mixing the seed with a MurmurHash mix (or any other quick bijective mixing function) before using it as a seed for PCG32 (maybe you're already doing something like that, in which case forget this comment).

Of course there will be always seeds leading to similar initialization if you use PCG32, but at least they will be random-looking.

Reply:
Taking the higher bits makes sense (this should also be done for the implementations in `rand_xoshiro`).

@vigna Would switching to splitmix for 64-bit seeding also avoid this problem?

Reply:
Yes, using SplitMix avoids visible dependence on the 64-bit seed.

Reply:
If there's an issue with our [`seed_from_u64`](https://github.com/rust-random/rand/blob/master/rand_core/src/lib.rs#L302) method, I'd prefer we deal with that directly.

Reply:
@dhardy I migrated `seed_from_u64` to splitmix64. This should work better with with similar seeds. However, this is a value-breaking change and requires bumping the minor version of `rand_core`.

Reply:
While @vigna does have a point, I'm tempted to ignore it: it's not very significant. I constructed a fake RNG with `[u32; 4]` state seeded from our `seed_from_u64` function, and this is what I get (with seeds `0, 1, 2, 1<<61, 1<<62, 1<<63`):
```
[0xf973f2ec, 0x45cdb581, 0x7346f087, 0xad6cad06]
[0x721dd8ea, 0x4e10265d, 0xf83b9c89, 0x2e78ce42]
[0xcc10d14d, 0x1e377cb1, 0x8e44efed, 0x07d77dee]
[0x3f2ec897, 0x145bdb58, 0x6f207734, 0x6ad2cad0]
[0xecd973f2, 0x8145adb5, 0x877346f3, 0x06acecad]
[0xf2ecb973, 0xb581458d, 0xf2877346, 0xad06ac6c]
```
So the first and last lines do have some similarities, its true, but are these correlations enough to cause noticeable correlation in Xoshiro128++ / 256++? Unlikely. Again, in binary:
```
0x0000000000000000: 11111001011100111111001011101100010001011100110110110101100000010111001101000110111100001000011110101101011011001010110100000110
0x0000000000000001: 01110010000111011101100011101010010011100001000000100110010111011111100000111011100111001000100100101110011110001100111001000010
0x0000000000000002: 11001100000100001101000101001101000111100011011101111100101100011000111001000100111011111110110100000111110101110111110111101110
0x2000000000000000: 00111111001011101100100010010111000101000101101111011011010110000110111100100000011101110011010001101010110100101100101011010000
0x4000000000000000: 11101100110110010111001111110010100000010100010110101101101101011000011101110011010001101111001100000110101011001110110010101101
0x8000000000000000: 11110010111011001011100101110011101101011000000101000101100011011111001010000111011100110100011010101101000001101010110001101100
```
And for SplitMix64:
```
0x0000000000000000: 01111011000111011100110110101111111000100010000010101000001110011010000110111001011001011111010001101110011110001001111001101010
0x0000000000000001: 10001001000000100101110011000001100100010000101000101101111011000110010110001110111011000110011110111110111010111000110110100001
0x0000000000000002: 00011100100101110101011011001110100101110101100000110101110111100000101111111100000111100100001010111111110010000100011000010000
0x2000000000000000: 00010100110100110000000000101011110011111011100010111101110000111111000111100000101010010001101011000010111001100110011110000100
0x4000000000000000: 10001110000011111010100111101011000000001010101001010000111010101010101100110010000101111111110110111101011001110110010010111111
0x8000000000000000: 00010010101010011111001111011011010010000001111011000000101000101010011000110000100100000001001011000100011011111010011000111000
```
Code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=88364b5a9665caecea34069a690bd660

We could just use a cryptographic hash function here; it's not performance critical, though we prefer to minimise code size. In the latter case, we already have [rand_seeder](https://github.com/rust-random/seeder/) and could make a cut-down version, but probably that's overkill.

Reply:
Yeah, it's a somewhat side concern, but it's five additional lines of code. Note that the examples with seed 1 and 2 are not part of the pattern—they have different lower bits. In binary you won't see the repeated patterns, but they are very visible in hex.

On the other hand, why don't you use another generator that's already in the library and does not have this problem?

Reply:
Because the `seed_from_u64` function is implemented in `rand_core` which is supposed to be a minimal bridge crate. It can be used without the rest of the library (at least in theory — half the original rationale for the split is no longer applicable now that `getrandom` is a completely separate library).

Reply:
Oh I see. So the issue propagates to every generator using that method. I would have used something with less dependence from the initial value, but I agree that it's unlikely significant problems might surface.

Reply:
@dhardy 
> I'm tempted to ignore it: it's not very significant.

What is the argument for continuing to use PCG? Avoiding churn? I think it would be nice to avoid issues such as #905, even if it does not matter that much in this case.

> We could just use a cryptographic hash function here; it's not performance critical, though we prefer to minimise code size.

Switching to splitmix is approximately the same number of lines and does not suffer from the correlation problem. As far as I can tell, the only downside is that this is a value-breaking change for everyone using `seed_from_u64`.

Reply:
@vks #905 is not an "issue" so much as incorrect documentation. To quote @vigna:

> Personally, I don't believe in the "middle ground" of "difficult-to-predict-but-not-crypto": usually it just means it is so easy to break that nobody tries. I mean, I can claim that xoshiro256++ is "difficult to predict"—and maybe that's even true. But until someone tries for real, that's just a vague and unsubstatiated claim. 

The main issue with PCG, as far as I am aware, is that streams tend to be quite similar; this is (part) of the issue in #1032 for example. This really doesn't matter for `seed_from_u64` since it only uses a single stream.

If we really want a perfect initialiser we should use a proven cryptographic PRNG or hash function, but to some extent this is overkill considering we only start with (a maximum of) 64-bits of entropy. It seems to me it is "good enough" for intended uses. If @vigna strongly disagrees, I will listen to his advice (though what I understood from the above is that we should use a hash function such as murmurhash *and* a PRNG).

We *also* are trying to keep the code-size of `rand_core` minimal however. I'm not entirely convinced the split between `rand` and `rand_core` still makes sense, but changing that now is more churn than justified in a mostly-mature library.

Between SplitMix64 and PCG32, I remain unconvinced that either is significantly better than the other. PCG does more mixing overall (since it emits only half the state per round, and does both multiplication and addition vs just addition), while SplitMix may have a stronger output function. Either way, I think getting good avalanche from low-order bits is much more important than from high-order bits due to the way users are likely to use this, and in this case I think both PRNGs are acceptable. But if @vigna knows better, please do correct me.

Reply:
> #905 is not an "issue" so much as incorrect documentation.

I was referring to the issue with similar sequences for different seeds.

> This really doesn't matter for seed_from_u64 since it only uses a single stream.

Yes, but a similar issue persists anyway: Different (but similar) seeds give very similar sequences, as discussed above.

> I'm not entirely convinced the split between rand and rand_core still makes sense, but changing that now is more churn than justified in a mostly-mature library.

I agree. For Rand 1.0 I could imagine merging the two crates and making `rand` without default features equivalent to the old `rand_core`.

> Between SplitMix64 and PCG32, I remain unconvinced that either is significantly better than the other. PCG does more mixing overall (since it emits only half the state per round, and does both multiplication and addition vs just addition), while SplitMix may have a stronger output function. Either way, I think getting good avalanche from low-order bits is much more important than from high-order bits due to the way users are likely to use this, and in this case I think both PRNGs are acceptable. But if @vigna knows better, please do correct me.

Alright, I will revert the splitmix commit. If necessary, we can perform this change in the future.

Reply:
One of the issues with PCG is the subsequences of the same generator are very similar, as shown above. You're confusing this issue with the issue that sequences from different streams are similar. The first one is obviously proven by the example you're showing. This is the issue at hand.

The persistent bit patterns of PCG32 for correlated seed cannot happen with SplitMix, so, yes, there is a significant and measurable difference. You can take two states of PCG32 and get highly correlated subsequences in which repeated bit patterns appear. You cannot do the same with SplitMix. You might be unconvinced, but it's mathematics. I'm not claiming that SplitMix will have uncorrelated subsequences, because the state mixing happens only leftward, as with PCG, but the problem is mitigated enormously by the powerful mix function.

If you're OK with repeated bit patterns in the initial state of your generators, that's another question—it's your library, if you want them, keep them.